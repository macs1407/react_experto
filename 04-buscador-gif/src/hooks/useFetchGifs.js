import { useState, useEffect } from "react"
import { getGifs } from "../helpers/getGifs";


// Un custom hook no es mas que una funcion
export const useFetchGifs = (categoria)=>{
    const [state, setState] = useState({
        data:[],
        loading:true,
    });

    // Se ejecuta cuando haya un cambio sobre una lista de dependencias si no se le envia nada en [] solo se ejecutara
    // Una vez cuando el componente se renderiza
    useEffect(() => {
        getGifs(categoria).then(imgs=>{
            setTimeout(()=>{
                setState({
                    data:imgs,
                    loading:false
                });
            },3000);
          
        });
    }, [categoria]);

    return state;
}