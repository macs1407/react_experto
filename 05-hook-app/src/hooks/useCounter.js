import { useState } from "react";

// Para crear un custom hook, hacemos lo siguiente
export const useCounter = (initialState = 10) => {
    // 1-) Establecer el useState
    const [counter, setCounter] = useState(initialState);

    const increment = ()=>{
        setCounter(counter + 1);
    }

    const decrement = ()=>{
        setCounter(counter - 1);
    }

    const reset = ()=>{
        setCounter(initialState);
    }
    // Lo que retorna el custom hook
    return {
        counter,
        increment,
        decrement,
        reset
    }
}
