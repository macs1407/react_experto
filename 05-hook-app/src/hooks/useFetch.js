import { useState, useEffect, useRef } from "react"

export const useFetch = (url) => {

    // Mantener la referencia a este hook cuando siga vivo o cuando el componente que lo usa esta montado
    const isMounted = useRef(true);

    const [state, setState] = useState({data: null, loading: true, error: null});

    // Solo se va a ejecutar una vez
    useEffect(() => {
        // y cuando se desmonte el useRef pasa a ser falso indicando que el componente ya no
        // esta montado
        return ()=>{
            isMounted.current = false;
        }
    }, []);

    useEffect(() => {
        setState({loading:true,error:null,data:null});

        fetch(url).then(resp=>resp.json())
                  .then(data=>{
                      setTimeout(()=>{
                        // Si el comoponente aun esta montado hace el setState, esto haca por problemas de fuga
                        // de memoria y errores al tratar de montar el componente
                        if(isMounted.current){
                            setState({
                                loading:false,
                                error:false,
                                data
                            });
                        } else {
                            console.log('setState no se llamo');
                        }
                      },3000);                    
                   });
    }, [url]);
    return state;
}
