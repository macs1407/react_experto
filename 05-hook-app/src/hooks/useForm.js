import { useState } from "react"

export const useForm = (initialState={}) => {
    const [values, setValues] = useState(initialState);

    const reset = ()=>{
        setValues(initialState);
    }

    const handleInputChange = (e)=>{
         // Se haca una copia del estado actual con ... o si no sobre escrbiriamos el state actual
        // el [target.name] hace referencia a name y email
        setValues({
            ...values,
            [e.target.name]: e.target.value
        });
    }

    return [values, handleInputChange, reset];
}
