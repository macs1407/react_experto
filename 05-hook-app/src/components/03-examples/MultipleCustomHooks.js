import React from 'react';
import '../02-useEfect/efects.css';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

export const MultipleCustomHooks = () => {

    const {counter, increment} = useCounter(1);
    const {loading, data} = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    // Con esta condicion le decimos que verifique que traiga informacion
    // del arreglo en la posicion 0
    const {author, quote} = !!data && data[0];

    return (
        <div>
            <h1>BreakingBad Quotes</h1>
            {
                loading 
                    ?
                        (
                            <div className="alert alert-info text-center">Loading....</div>
                        )
                    :
                        (
                            <blockquote className="blockquote text-right">
                                <p className="mn-0">{quote}</p>
                                <footer className="blockquote-footer">{author}</footer>
                            </blockquote>
                        )
            }
            
            <button className="btn btn-primary" onClick={increment}>Siguiente frase</button>
        </div>
    )
}
