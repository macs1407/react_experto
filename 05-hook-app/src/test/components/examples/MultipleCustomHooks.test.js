import React from 'react';
import '@testing-library/jest-dom';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import {shallow} from 'enzyme';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas con MultipleCustomHooks', () => {

    useCounter.mockReturnValue({
        counter:10,
        increment:()=>{}
    });

    test('Debe de levantar MultipleCustomHooks ', () => {
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        });
        const wraper = shallow(<MultipleCustomHooks />);
        expect(wraper).toMatchSnapshot();
    });
    
    test('Debe de mostrar la informacion ', () => {
        useFetch.mockReturnValue({
            data:[{
                author:'maikol',
                quote:'hola'
            }],
            loading:false,
            error:null
        });
        const wraper = shallow(<MultipleCustomHooks />);
        expect(wraper.find('.alert').exists()).toBe(false);
        expect(wraper.find('.mn-0').text().trim()).toBe('hola');
    });
})
