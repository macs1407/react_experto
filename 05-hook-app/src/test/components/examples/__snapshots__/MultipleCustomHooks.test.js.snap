// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con MultipleCustomHooks Debe de levantar MultipleCustomHooks  1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading....
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
